### PR TITLE
Add Either.filterOrLeft

### PR DESCRIPTION
**Pull request by extremegf**
Hello, 

Here's a proposal for `Either.filterOrLeft` based on `Effect.filterOrFail`. This is particularly useful in non-effectful pipelines, to validate the produced result. I considered names `filterOrFail`, `filterOrElse`, `validate`, `validateWith` - I'm open to suggestions.

 Also I don't know how to handle `@since `

### CODE REVIEW

File: src/Either.ts [hunk 1]
```diff
@@ -375,6 +376,59 @@ export const match: {
...
+ * )
+ * assert.deepStrictEqual(
+ *   pipe(
+ *     E.right(0),
+ *     E.filterOrLeft(isPositive, n => `${n} is not positive`)
+ *   ),
+ *   E.left("0 is not positive")
+ * )
+ *
+ * @since TODO(horban): add since
```
Comment (tim-smart) [inline src/Either.ts:404]: You can use 2.0.0 here. This represents the major version of when the api was introduced.

**Review by tim-smart — CHANGES_REQUESTED** [review]

**Comment by gcanti** [comment]
If we are going to add this, I vote for `filterOrFail`

**Comment by gcanti** [comment]
yeah, if we are going to add back some biased APIs I would slightly prefer consistency with Effect, though I guess filterOrLeft is also acceptable since there's already mapLeft

File: src/Either.ts (hunk 1 above)
Comment (extremegf) [reply src/Either.ts]: Done.

**Comment by tim-smart** [comment]
My vote is with Left, to keep consistency with the other APIs in the module.

**Review by tim-smart — APPROVED** [review]

**Comment by extremegf** [comment]
Let me know if you have some small stuff to take care of.

**Pull request merged**